File: pluginapk/src/main/java/com/malin/plugin/BaseActivity.java
package com.malin.plugin;

import android.content.Context;
import android.content.res.Resources;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ContextThemeWrapper;

import java.io.File;
import java.lang.reflect.Field;

public class BaseActivity extends AppCompatActivity {

    protected Context mContext;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Resources resource = PluginResourceUtil.getResource(getApplication());
        mContext = new ContextThemeWrapper(getBaseContext(),  R.style.AppCompatThemePlugin);
        Class<?> contextClazz = mContext.getClass();
        try {
            // android 15/16 java.lang.NoSuchFieldException: mResources
            // fix bug
            // use androidx.appcompat.view.ContextThemeWrapper; Instead of android.view.ContextThemeWrapper;
            Field mResourcesField = contextClazz.getDeclaredField("mResources");
            mResourcesField.setAccessible(true);
            mResourcesField.set(mContext, resource);
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }
}

File: pluginapk/src/main/java/com/malin/plugin/BaseActivity.kt
package com.malin.plugin

import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ContextThemeWrapper

open class BaseActivity : AppCompatActivity() {

    protected var mContext: Context? = null
    protected var pluginInHostRunning = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        var hostAppClazz: Class<*>? = null
        try {
            hostAppClazz = Class.forName("com.malin.hook.MApplication")
        } catch (ignore: Throwable) {
        }
        pluginInHostRunning = hostAppClazz != null
        if (pluginInHostRunning) {
            // 插件apk在宿主中运行
            val resource = PluginResourceUtil.getResource(application)
            mContext = ContextThemeWrapper(baseContext, 0)
            val contextClazz: Class<*> = mContext?.javaClass ?: return
            try {
                val mResourcesField = contextClazz.getDeclaredField("mResources")
                mResourcesField.isAccessible = true
                mResourcesField[mContext] = resource
                val rClazz = Class.forName("com.google.android.material.R\$style")
                val themeField = rClazz.getDeclaredField("Theme_MaterialComponents_DayNight")
                themeField.isAccessible = true
                val themeObj = themeField[null]
                if (themeObj != null) {
                    val theme = themeObj as Int
                    val mThemeResourceField = contextClazz.getDeclaredField("mThemeResource")
                    mThemeResourceField.isAccessible = true
                    mThemeResourceField[mContext] = theme
                }
            } catch (e: Throwable) {
                e.printStackTrace()
            }
        } else {
            // 插件作为独立的apk运行
            mContext = null
        }
    }
}
